Add kapacitor state and execution module

anlutro opened the pull request:
Kapacitor is a monitoring/alerting system for InfluxDB.

There's no hurry in merging this branch. Criticism and code review welcome.

Follow-up to #31716

anlutro committed:
add kapacitor state and module

anlutro committed:
lint fixes

anlutro committed:
tweak behaviour and add tests for kapacitor

rallytime commented on salt/modules/kapacitor.py:
```diff
@@ -0,0 +1,147 @@
+    This data can also be passed into pillar. Options passed into opts will
+    overwrite options passed into pillar.
+
+.. versionadded:: 2016.3.0
```
Can you change `2016.3.0` to `Carbon`?

rallytime commented on salt/modules/kapacitor.py:
```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+'''
+Kapacitor execution module.
+
```
Seems obvious, but let's add a `:depends: kapacitor` declaration here.

anlutro replied:
As long as a correct API endpoint (host + port) is configured in the minion config, nothing needs to be installed on the Salt minion - not even Kapacitor itself. Do we still need the `depends` block? If so, maybe be a bit more specific so people don't go looking for a python module named kapacitor?

rallytime replied:
Ah, that's my misunderstanding of what kapacitor is. In that case, the `depends` block isn't needed.

rallytime commented on salt/states/kapacitor.py:
```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+'''
+Kapacitor state module.
+
```
Same comment/question about adding dependency notes here as well.

rallytime commented on salt/states/kapacitor.py:
```diff
@@ -0,0 +1,161 @@
+    This data can also be passed into pillar. Options passed into opts will
+    overwrite options passed into pillar.
+
+.. versionadded:: 2016.3.0
```
Should be `Carbon` now.

rallytime commented on salt/states/kapacitor.py:
```diff
@@ -0,0 +1,161 @@
+    changes = []
+    ret = {'name': name, 'changes': {}, 'result': True, 'comment': ''}
+
+    if not tick_script:
```
If the `tick_script` must be provided, then why not make that a required arg instead of a kwarg and avoid this check?

anlutro replied:
Somehow, I got the impression that all required args need to be kwargs to work with Salt's state system. Clearly a misconception!

rallytime replied:
Right before I wrote that comment I was thinking along those same lines, too, but I found some examples where that isn't necessarily the case. The `file.replace` and `file.line` (which are both heavily used) both have args and kwargs.

rallytime commented on salt/states/kapacitor.py:
```diff
@@ -0,0 +1,161 @@
+        if task and task['Enabled']:
+            comments.append('Task is already enabled')
+        else:
+            if __opts__['test']:
```
When writing states, let's be very careful about when setting the `result` to `None` when passing `test=True`. If we're returning `None` in the result, that indicates that when you run your states with `test=true`, then `changes` *will* be made. So what should be done here is to track what things will be changing in this state. This is useful both for `test=true` functionality as well as seeing an output of what happened/changed at the end of a successful state run. The `changes` dictionary should be tracking this more than the `comments` key.

The `changes` dictionary should have an `old` and a `new` key to indicate what is changing. A state should keep track of multiple pieces within it's changes dict. For example, it might be useful to have something like
```
{'changes': {'task script': {'old': '', 'new': task_script}, 'enabled': {'old': 'disabled', 'new': 'enabled'}}}
```
In the results of the state run. Then, once you have a populated list of (potential) changes you can see if `test=true` was passed in, and set result to `None`.

anlutro replied:
I'll work on this - but for the TICKscript, it can get very verbose to render both the old and new version. That's why I used difflib instead. Would something like this be fine?

```
changes: {
  'tick script': DIFF_HERE,
  'enabled': {
    'old': False,
    'new': True
  }
}
```

rallytime replied:
Yeah, using a diff sounds good. I know we use diffs in other changes dictionaries so you could use that as an example. Check out the file state module for some examples on how they're doing that, if you like.

rallytime commented:
Hi @anlutro - Thanks for moving this to the develop branch. This is great! There are some [pylint errors](https://jenkins.saltstack.com/job/salt-pr-lint-n/14087/violations/) that need to be cleaned up. 

I have also left a couple of comments, but my last comment is the most important. We've had some troubles in the past with other states where the changes dictionary isn't properly populated or considered and `test=true` functionality always returns `None`, even if no changes will occur. So, I am being picky here and trying to avoid those types of bugs at the pass. If you could take a look at refactoring your states here to make sure the changes dictionary is populated correctly when changes are expected as well as ensuring we're checking for that `test=true` result of `None` correctly, that would be fantastic.

anlutro committed:
correct versionadded

anlutro committed:
just make tick_script arg required

anlutro committed:
add comment when script is to be updated in test mode

anlutro committed:
switch changes-enabled to show old/new

anlutro committed:
rename key for diff in changes dict

anlutro commented:
I'm looking over at the return values, and I can't see that I'm setting result to `None` anywhere I shouldn't. The changes dictionary is also always populated when a change is going to get made. I've been bitten by problems with test mode myself in the past, so I'm well aware of the problems that can arise.

I've renamed the "diff" changes to "TICKscript diff", and changed the "enabled" changes to show old and new, so the changes dict should be a bit easier to read. If there's more you'd like me to look over in terms of what the state result/change is, you'll have to be specific.

anlutro committed:
update tests with new changes structure

rallytime commented:
@anlutro Awesome! Thanks for reviewing those. This looks great. There's only one [tiny lint error](https://jenkins.saltstack.com/job/salt-pr-lint-n/14113/violations/file/salt/states/kapacitor.py/) left, and then we'll get this in.

rallytime commented:
I grabbed the lint failure really quickly. Thanks again for your work!

rallytime commented:
@anlutro No problem! It's a little tricky to spot what's going on here as we are doing some upgrades on our PR tests.
